### PR TITLE
Speed up ordinal lookups in composite aggregation

### PR DESCRIPTION
**Pull request by jimczi**
This change is an optimization on top of #74559, that sorts ordinals to perform
lookups. The sorting ensures that we don't do the de-compression of
 blocks in the dictionary of terms more than necessary.
 In the worst case today, we can decompress the same block for each lookup
term per segment, while this change requires only one decompression.

This commit also creates the doc values lookup once
per request per segment. This is useful when inverted lists
 are used to shortcut the collection since terms are already sorted
 in the dictionary.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java:315]: should this just return `next` now that the collect method is just delegating?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java:365]: `lastOldOrd` is not set anywhere (so this optimization is effectively disabled :D). Also, should we only reuse the previous ord when `oldOrd >= 0`? In case where it's negative, we might have different values in valuesUnmapped.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java:339]: In case we have two negative ords here that are equal, should we fall back to a comparison of unmapped values in order to bring them in order? Otherwise, in case of super high cardinality fields, we might still do random access.

**Review by ywelsch — COMMENTED** [review]
Thank you for looking into this. I've left two comments to address (and a bunch of optional ones)

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java]: Addressed in https://github.com/elastic/elasticsearch/pull/78313/commits/18ee058c3e2862aa2f99f732e27b53a423ced38b

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java]: I pushed https://github.com/elastic/elasticsearch/pull/78313/commits/18ee058c3e2862aa2f99f732e27b53a423ced38b

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java]: Good catch, I pushed https://github.com/elastic/elasticsearch/pull/78313/commits/18ee058c3e2862aa2f99f732e27b53a423ced38b and also updated the unmapped value when a duplicate is found

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java:348]: Good idea, I pushed https://github.com/elastic/elasticsearch/pull/78313/commits/18ee058c3e2862aa2f99f732e27b53a423ced38b

**Comment by not-napoleon** [comment]
Have we run the ML performance test against this branch?  This seems like it should help, but it'd be good to have the numbers to back that up.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java:306]: This comment isn't right for this method. Here, after your refactoring, we don't have a separate instance as there is no iteration anymore.

**Review by ywelsch — APPROVED** [review]
I've left two more minor comments, looking good o.w.

**Review by imotov — COMMENTED** [review]
> Have we run the ML performance test against this branch? This seems like it should help, but it'd be good to have the numbers to back that up.

It would be great to have the results of this test before we merge this in, otherwise without an existing performance infrastructure, I feel like we are running blind with these changes.

**Comment by ywelsch** [comment]
>> Have we run the ML performance test against this branch? This seems like it should help, but it'd be good to have the numbers to back that up.

> It would be great to have the results of this test before we merge this in, otherwise without an existing performance infrastructure, I feel like we are running blind with these changes.

Jim has done some ad-hoc testing. 

Once this PR is merged (e.g. master-only in a first step), it can be validated by the ML QA team (@wwang500).

There's ongoing work to add benchmarking capabilities to Rally for paging through composite aggregations.

I would like this PR to go into 7.16 and not be blocked on benchmarking infrastructure.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java:376]: The slots are per-document so we can have duplicates (i.e.: when more than one source is provided).

**Pull request merged**
